hamilton/lifecycle/caching.py

```diff
@@ -0,0 +1,1087 @@
+
+    # subclass `BaseDoNodeExecute` instead of `NodeExecutionMethod` to access `node.Node` object
+    def do_node_execute(
+        self,
+        *,
+        node_: hamilton.node.Node,
+        kwargs: Dict[str, Any],
+        task_id: Optional[str] = None,
+        **future_kwargs,
+    ):
+        """If the metadata store includes the context key, else compute result.
+        If a fingerprint is retrieved, but reading from cache fails, invalidate the metadata store
+        and force recompute and storage of fingerprint + result.
+        """
+        # gather the attributes typically available through NodeExecutionMethod
+        node_name = node_.name
+        node_callable = node_.callable
+        node_kwargs = kwargs
+
+        if node_.node_role == hamilton.node.NodeType.EXPAND:
+            self._expand_nodes.add(node_name)
+
+        elif node_.node_role == hamilton.node.NodeType.COLLECT:
+            self._collect_nodes.add(node_name)
+            collected_name = node_.collect_dependency
+            collected_values = node_kwargs[node_.collect_dependency]
+            # Counter is the right data structure: it's unordered, arbitrary length, allow duplicates
+            data_version_counter = collections.Counter(collected_values)
+            collected_data_version = fingerprinting.hash_mapping(dict(data_version_counter))
+            # TODO apply the right caching behavior; usually applied in `run_after_node_execution`
+            self.set_memory_metadata(
+                node_name=collected_name, task_id=task_id, data_version=collected_data_version
+            )
+
+        # node "inside" a parallel block
+        elif task_id != node_name:
+            for dep_name, dep_value in node_kwargs.items():
+                # version single item yielded by a `Parallelizable` node
+                if dep_name in self._expand_nodes:
+                    # use the hook manually to correctly apply caching behavior.
+                    self.run_after_node_execution(
+                        node_name=dep_name, task_id=task_id, result=dep_value
+                    )
+
+        node_behavior = self.behaviors[node_name]
+        if node_behavior in (CachingBehavior.DISABLE, CachingBehavior.RECOMPUTE):
+            result = self.execute_node(
+                node_name=node_name,
+                node_callable=node_callable,
+                node_kwargs=node_kwargs,
+                task_id=task_id,
+            )
+            return result
+
+        # create context_key before execution; will be reused after execution
+        context_key = self.create_context_key(
+            node_name=node_name, dep_names=node_kwargs.keys(), task_id=task_id
+        )
+        self.set_context_key(node_name=node_name, task_id=task_id, context_key=context_key)
+
+        # check in-memory metadata
+        data_version = self.get_memory_metadata(node_name=node_name, task_id=task_id)
+
+        # check stored metadata
```

Comment:
Design question -- why are we separating stored from in-memory metadata? When do we sync them?

Reply:
While this is a convoluted, the goal was to limit round-trips to the result store. The physical metadata store is always a super set of what's in memory. 

Invalidating the physical store is a way to force recompute. This should be more explicit as mentioned in another comment

Reply:
👍